ISPN-11723 Cluster Backup/Restore tool

ryanemerson opened the pull request:
https://issues.redhat.com/browse/ISPN-11723
https://issues.redhat.com/browse/ISPN-12095

CLI support to follow in a separate PR [ISPN-12095](https://issues.redhat.com/browse/ISPN-12095).

There's still some polishing required, mostly centred around threads in the `BackupWriter` and `BackupReader`.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/BackupManager.java:
```diff
@@ -0,0 +1,31 @@
+    * @param backup the bytes of the uploaded backup file.
+    * @return a {@link CompletionStage} that completes when all of the entries in the backup have been restored.
+    */
+   CompletionStage<Void> restore(byte[] backup);
```
`@param backup the bytes of the uploaded backup file.` <= what is this exactly? is the full data in a single byte{]?

tristantarrant replied:
This should really be an InputStream or a ReadableByteChannel

ryanemerson replied:
Sorry I forgot back to get to your original comment Pedro. 

Agreed it shouldn't be byte[], I will update this in the interface. We do have a bigger issue at the moment though, which is AFAIK we can only receive the bytes from a `RestRequest` via `ContentSource#rawContent()` which returns a `byte[]`. As backups will most likely be large, we should support Streams for uploads/downloads with HTTP2, maybe with a fallback of chunked transfer encoding for HTTP/1.1

\cc @gustavonalle

gustavonalle replied:
Do we really need to send a giant amount of data to the server?  Can't you just accept a ```Path``` to a file somewhere?

If we really need it, it's a great opportunity to expand the REST framework to support a stream of data, using chunked transfer, as it'd be supported both for HTTP/1 and HTTP/2

ryanemerson replied:
> Do we really need to send a giant amount of data to the server? Can't you just accept a `Path` to a file somewhere?

How do you get the file from the various clients to the server if you don't upload it? The idea is that this will work in Openshift and can be triggered via the Cli and eventually the console.

We could also add the option of the server importing the contents of a local backup file on server startup. In which case the server can pass a `InputStream` of the local file to the `restore` method.

ryanemerson replied:
> If we really need it, it's a great opportunity to expand the REST framework to support a stream of data, using chunked transfer, as it'd be supported both for HTTP/1 and HTTP/2

I'm not sure if you saw my earlier comment, but my understanding is that chunked transfer was removed in `HTTP/2` https://en.wikipedia.org/wiki/Chunked_transfer_encoding.

gustavonalle replied:
Our REST server is using HTTP/1 tunnelled in HTTP/2 frames, so it should work.

gustavonalle replied:
> How do you get the file from the various clients to the server if you don't upload it?

I was just wondering 1 TB backups being sent through HTTP from the client to the server and got a bit worried :)

gustavonalle replied:
One question: the other way around, i.e., producing the backup, does it first create it on disk and then send it to the client? If so, maybe chunking is not the best but https://en.wikipedia.org/wiki/Byte_serving

ryanemerson replied:
> does it first create it on disk and then send it to the client

Yep. Currently this is a very simple implementation where it's stored as `staging.zip`, however if we take advantange of byte serving and assign an Id per backup request, we could allow for failed downloads to be resumed. Although we would probably need some form of garbage collection for backups whom's downloads never complete.

gustavonalle replied:
ok, we can always revisit it later

ryanemerson replied:
I've created https://issues.redhat.com/browse/ISPN-12145 to track the implementation of chunked transfer encoding.

ryanemerson commented on core/src/main/java/org/infinispan/util/concurrent/BlockingManager.java:
```diff
@@ -106,6 +106,21 @@
 
+   /**
+    * Replacement for {@link CompletionStage#thenRunAsync(Runnable)} that invokes the {@code Runnable} in a blocking thread
+    * (if the current thread is non-blocking) or in the current thread (if the current thread is blocking).
```
I definitely think this is a fix for a different PR. In this case I just used `thenApplyBlocking` as a basis of this implementation, so I assume that also has the same issue.

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/BackupReader.java:
```diff
@@ -0,0 +1,140 @@
+   CompletionStage<Void> restore(InputStream is, Map<String, BackupManager.ContainerResources> params) {
+      final Path stagingFile = rootDir.resolve(STAGING_ZIP);
+
+      CompletionStage<Void> createStagingFile = blockingManager.runBlocking(() -> {
```
I was trying to make it more fine-grained, but after your explanation I can see that doesn't really make sense as the operations obviously depend on each other :slightly_smiling_face:

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/BackupReader.java:
```diff
@@ -0,0 +1,140 @@
+         }
+      }, "read-manifest");
+
+      return blockingManager.thenRun(processContainers, () -> {
```
I have updated it to `return processContainers.thenRun(() -> {...});` as we only require `CompletionStage<Void>`.

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/BackupWriter.java:
```diff
@@ -0,0 +1,177 @@
+   }
+
+   CompletionStage<Path> create(Map<String, BackupManager.ContainerResources> params) {
+      List<CompletionStage<?>> stages = new ArrayList<>(params.size() + 1);
```
I completely forgot that existed :slightly_smiling_face:

ryanemerson replied:
I have updated all instances that were previously using `CompletionStages#allOf(Collection)` and removed that static method.

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java:
```diff
@@ -0,0 +1,117 @@
+      this.cacheManagers = cacheManagers;
+      this.reader = new BackupReader(blockingManager, cacheManagers, rootDir);
+      this.writer = new BackupWriter(blockingManager, cacheManagers, rootDir);
+      rootDir.toFile().mkdir();
```
Good catch. As the `rootDir` is provided via `Server.java` I have moved the creation there.

gustavonalle commented on server/tests/src/test/java/org/infinispan/server/functional/ClusterBackupIT.java:
```diff
@@ -0,0 +1,186 @@
+import org.junit.Test;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
```
Jackson databind is being killed, please avoid using it and use the internal mJson instead

gustavonalle commented on server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:
```diff
@@ -1,33 +1,52 @@
       return new Invocations.Builder()
             .invocation().methods(POST).path("/v2/cluster").withAction("stop").handleWith(this::stop)
+            .invocation().methods(GET).path("/v2/cluster").withAction("backup").handleWith(this::backup)
+            .invocation().methods(POST).path("/v2/cluster").withAction("restore").handleWith(this::restore)
```
Please add documentation in the REST API for those two methods

ryanemerson commented:
I have updated the PR so that the following REST endpoints are exposed in order to facilitate more context aware operations in the CLI. I have also removed the CLI commit from this PR for now, as I am still reworking this aspect and this PR is already ~ 3k lines.

- /v2/cluster?action=backup exposes backup of all containers and content
- /v2/cluster?action=restore restores all containers and content from
      uploaded zip file
- /v2/cache-managers/{name}?action=backup exposes backup of all
      container content
- /v2/cache-managers/{name}?action=backup&<resource>=<resource names|*>
      exposes parameterised backup
- /v2/cache-managers/{name}?action=restore exposes full and
      parameterised backup

I have also update the `BackupManagerImpl` to utilise `ClusteredLock` instances to ensure that only a single backup/restore operation can occur at any one time.

ryanemerson commented on documentation/src/main/asciidoc/topics/ref_rest_cache_managers.adoc:
```diff
@@ -204,3 +204,112 @@ include::json_examples/rest_cache_stats_response.json[]
+
+[source,options="nowrap",subs=attributes+]
+----
+GET /rest/v2/cache-managers/{cacheManagerName}?action=backup
```
@gustavonalle As a backup is non-idempotent does it make more sense for this to be a POST?

gustavonalle replied:
Does it trigger a backup or simply re-downloads something that was already produced?

ryanemerson replied:
It always initiates a new backup and downloads the result.

gustavonalle replied:
According to the newly recent changes we were required to do, everything with side-effect should be a POST. 

But a POST downloading something is not "RESTful", and according to [1], there is a way to handle this: return a 201 code with a location header of the newly created resource or 202 if the resource was accepted but not ready.

[1] https://tools.ietf.org/html/rfc7231#section-6.3.2


I think we should make ```backup``` a proper resource:

1) Trigger the backup with a certain name and return ```202``` (Accepted)
```
POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

2)  Get the backup, returning it with ```200``` or ```202``` if it's not ready yet
```
GET /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

3) Delete a certain backup
```
DELETE /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

4) Restore a backup
```
POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}?action=restore
```

ryanemerson replied:
I like this proposal a lot :+1:

For the CLI I can make it so that stages 1..3 appear as a single operation

ryanemerson replied:
> 4. Restore a backup
> 
> ```
> POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}?action=restore
> ```

For this I have added very basic support for `multipart/form-data` that way the parameters used to filter the content from the backup file can be included as part of the payload.

gustavonalle replied:
If I am not mistaken, we haven't been using ```multipart/form-data```, instead just use the JSON document as payload. 
FYI, this is how search handles it:

```POST /rest/v2/caches/{cacheName}?action=search```
```json
{
 "query":"from Entity where name:\"user1\"",
 "max_results":20,
 "offset":10
}
```

query params are part of the document

ryanemerson replied:
The difference here is that we need to send both the `application/json` and the `application/zip` in a single request.

gustavonalle replied:
fair enough, I was imagining that backups wouldn't travel via HTTP, but rather be triggered, restored and deleted from a known folder only.

ryanemerson replied:
Unfortunately I don't believe that approach is flexible enough for Openshift and restoring backups to rolled out clusters. We [previously discussed](https://github.com/infinispan/infinispan/pull/8532#pullrequestreview-446504711) this and ways we can improve performance going forward, but I think for the initial PR at least the important thing to focus on is the REST API and BackupManager internals.

gustavonalle replied:
I am thinking the scenario for large datasets, say GBs (or TBs) of data. 

Starting the CLI from another continent pointing to the cluster, and asking a backup will be very slow. A restore would involve transferring that amount of data again, all via http.
If the CLI is run in the same network as the cluster, it wouldn't be so bad, but it would still be faster if the server writes the backup to a certain file, and restore from that same place, without the data passing through REST. 

Another thing related to sizes, the rest server where the backup is triggered need to have space to hold the full cluster worth of data, otherwise it will have disk errors. The way I imagine backups working on cloud env is to attach a larger volume, with a certain proximity to the cluster, so this would imply the possibility for the user to choose the directory of destination.

WDYT?

ryanemerson replied:
> The way I imagine backups working on cloud env is to attach a larger volume, with a certain proximity to the cluster, so this would imply the possibility for the user to choose the directory of destination.

I agree, this is the way to go. For typical Infinispan deployments there is no need for a large volume, so it's too restrictive to assume that  a server's volume would have the capacity to store content spread across the cluster.

I think that we should allow the following:

1. Allow users to specify an optional path when creating a backup. This determines both the working-dir and final location of the created backup file. If no path is specified, then we utilise the current location `ISPN_HOME/server/data/backup-manager/{backupName}/{backupName}.zip`

1. Restore via a path specified in the JSON payload. The file is processed from the locally available path with no file upload etc.

1. Restore via upload. Allows smaller backups to be restored without attaching additional volumes and simple Console user experience, i.e. upload backup via browser form.
    1. Users can optionally provide a custom working-dir path via JSON, if a larger volume is required for processing.

gustavonalle replied:
+1 for items 1), 2)

Still in doubt about passing backups through HTTP though, but I am not ferociously against it :smile:

So assuming we decide to add later the path reference based backups as described above (in a follow up PR), it'd be wise to avoid disruptions to the REST API. I suppose
```
POST /rest/v2/cache-managers/{cacheManagerName}/backups/{backupName}
```

Can be used for both cases right? If the backup is not there, then it should have a reference for a path.

ryanemerson replied:
> So assuming we decide to add later the path reference based backups as described above (in a follow up PR), it'd be wise to avoid disruptions to the REST API. I suppose

I'm actually making the changes atm and it doesn't alter things too much, so it should be Ok to make it into this PR.

> Can be used for both cases right? If the backup is not there, then it should have a reference for a path.

:+1: I was thinking the JSON would be the following for referenced based backup and restore requests:

```json
{
  "backup-dir": "/some/path",
  "resources": {
     "caches": "[*]"
   }
}
```
The Content-Type of the restore request determines what course of action is taken:
* `application/zip` uploads the backup and processes all resource
* `application/json` reference based restore, reading backups from `backup-dir/{backupName}.zip` and only processing resources in the "resources" object if specified.
* `multipart/form-data` uploads the backup and processes the specified resources in the json. "backup-dir" still respected as the working directory if specified.

ryanemerson replied:
Updated. I removed the option to upload via `application/zip` as netty allows files uploaded via `multipart/form-data` to be stored to disk, whereas that does not seem possible with `application/zip`. The latter is all that is required by html forms, so this shouldn't be an issue for the console.

ryanemerson commented:
Any and all reviews welcome :smile:. I will tidy up the git history further down the line.

ryanemerson commented:
None of the failures are related

gustavonalle commented:
The REST related code looks fine to me!

gustavonalle commented on server/rest/src/main/java/org/infinispan/rest/resources/CacheManagerResource.java:
```diff
@@ -108,6 +111,11 @@ public Invocations getInvocations() {
+            // BackupManager
+            .invocation().methods(GET).path("/v2/cache-managers/{name}/backups").handleWith(this::getAllBackupNames)
+            .invocation().methods(DELETE, GET, HEAD, POST).path("/v2/cache-managers/{name}/backups/{backupName}").handleWith(this::backup)
+            .invocation().methods(POST).path("/v2/cache-managers/{name}/backups").withAction("restore").handleWith(this::restore)
```
The REST related changes LGTM!

oraNod commented on client/rest-client/src/main/java/org/infinispan/client/rest/RestCacheManagerClient.java:
```diff
@@ -46,4 +49,74 @@
+   CompletionStage<RestResponse> createBackup(String name, String workingDir, Map<String, List<String>> resources);
+
+   /**
+    * Retrieve a backup file with the given name from the server.
```
```suggestion
    * Retrieves a backup file with the given name from the server.
```

oraNod commented on client/rest-client/src/main/java/org/infinispan/client/rest/RestCacheManagerClient.java:
```diff
@@ -46,4 +49,74 @@
+   CompletionStage<RestResponse> getBackup(String name, boolean skipBody);
+
+   /**
+    * Delete a backup file from the server.
```
```suggestion
    * Deletes a backup file from the server.
```

oraNod commented on client/rest-client/src/main/java/org/infinispan/client/rest/RestClusterClient.java:
```diff
@@ -17,4 +18,41 @@
+   CompletionStage<RestResponse> createBackup(String name);
+
+   /**
+    * Retrieve a backup file with the given name from the server.
```
```suggestion
    * Retrieves a backup file with the given name from the server.
```

oraNod commented on client/rest-client/src/main/java/org/infinispan/client/rest/RestClusterClient.java:
```diff
@@ -17,4 +18,41 @@
+   CompletionStage<RestResponse> getBackup(String name, boolean skipBody);
+
+   /**
+    * Delete a backup file from the server.
```
```suggestion
    * Deletes a backup file from the server.
```

oraNod commented on documentation/src/main/asciidoc/topics/backup_resources.adoc:
```diff
@@ -0,0 +1,25 @@
+.Resource Parameters
```
@ryanemerson hey, why is this in a separate adoc file?

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,282 @@
+   @Override
+   public CompletionStage<Void> backup() {
+      AggregateCompletionStage<Void> stages = CompletionStages.aggregateCompletionStage();
+      for (String cache : resources)
```
It makes me wonder if we should add some extra methods to `BlockingManager#BlockingExecutor` to handle these types of cases.

ryanemerson replied:
Agreed that we should limit the number of blocking threads are used by this, as even if there is a reasonably modest number of caches this could cause issues.

I think `BlockingManager.BlockingExecutor` should provide all of the `BlockingManager` methods with the exception of `limitedBlockingExecutor`. That way we can do something like:

```java
      BlockingManager.BlockingExecutor executor = blockingManager.limitedBlockingExecutor("cache-backup-executor", 5);
     for (String cache : resources)
         stages.dependsOn(createCacheBackup(cache, executor));
```

To reduce repetition in the `BlockingManager` interface, we can create another interface `BlockingOperations` and then do the following:

```java
public interface BlockingManager extends BlockingOperations {

    BlockingExecutor limitedBlockingExecutor(String name, int concurrency);

    interface BlockingExecutor extends BlockingOperations {
    }
}
```

`BlockingManagerImpl` would then just provide versions of the `BlockingOperations` methods that take a `Scheduler|Executor` as an additional parameter.

ryanemerson replied:
Similarly we could avoid `BlockingOperations` and add a method that returns a BackupManager implementations that utilies a `LimitedExecutor`:

```java
BlockingManager limitedBlockingManager(String name, int concurrency);
```

wburns replied:
I look forward to seeing this change in another PR 👍

ryanemerson replied:
I have created [ISPN-12226](https://issues.redhat.com/browse/ISPN-12226) to track the `BlockingManager` enhancements and [ISPN-12227](https://issues.redhat.com/browse/ISPN-12227) for the changes to the `BackupManager`.

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,282 @@
+         String dataFileName = dataFile(cacheName);
+         Path datFile = cacheRoot.resolve(dataFileName);
+
+         Publisher<CacheEntry<?, ?>> p = s -> clusterPublisherManager.entryPublisher(null, null, null, false,
```
This is something that I have been in two minds about. Originally I was thinking that if data is already persisted, then the stores can also be migrated by the user, but this isn't very user-friendly especially if stores are not shared and we have a large cluster. Having a centralised backup greatly simplifies this, so including stores by default makes more sense on second thought.

In the case of a large shared store, probably JDBC based, it's probably not desired to have the store content in the backup. Once the core backup/restore pieces are in place (CLI, Operator etc) we can add a parameter to toggle this behaviour.

wburns replied:
Makes sense. This makes me wonder if we should have the default value be if the store it uses is shared or not.

Also I wonder if we should add support for the Flag `SKIP_SHARED_CACHE_STORE` in the `ClusterPublisherManager`

ryanemerson replied:
> This makes me wonder if we should have the default value be if the store it uses is shared or not

Tbh I think this could be difficult to explain to the user in a transparent way. Having a consistent default and then documenting that it's possible to backup in-memory contents only is much simpler IMO. We can then add a note explaining that for larger shared stores it's probably desirable to enable this option.

ryanemerson replied:
> Also I wonder if we should add support for the Flag `SKIP_SHARED_CACHE_STORE` in the `ClusterPublisherManager`

Makes sense. It would be perfect for this use-case

ryanemerson commented on client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestCacheManagerClientOkHttp.java:
```diff
@@ -121,4 +128,75 @@ public String name() {
+
+      if (resources != null) {
+         Json resourcesJson = Json.object();
+         resources.forEach((k, v) -> resourcesJson.set(k, v.toArray(new String[0])));
```
> Do we need to escape any json characters here? Could the String contain valid json delimiters?

I don't think so, there shouldn't be any JSON passed in the `Map`.

> Also does Json.factory().make(resources) not do what we want?

Yes :slightly_smiling_face:. I didn't know about that before.

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,282 @@
+                              be.lastUsed = e.getLastUsed();
+                              return be;
+                           })
+                           .doOnNext(e -> writeMessageStream(e, serCtx, output))
```
I don't think so, because `subscribe` returns a `Disposable` but we need to return a `Publisher` so that it can be consumed as the `Flowable.using` `sourceSupplier`. This is similar to the `JpaStore#createBatchFlowable` code.

wburns replied:
👍 Forgot it returns a Disposable :(

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,282 @@
+               () -> Files.newOutputStream(datFile),
+               output ->
+                     Flowable.fromPublisher(p)
+                           .buffer(BUFFER_SIZE)
```
I honestly can't remember. Removed.

ryanemerson commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,282 @@
+   }
+
+   private CompletionStage<Void> createCacheBackup(String cacheName) {
+      return blockingManager.runBlocking(() -> {
```
I'm not sure what you mean here. I have updated `createCacheBackup` to use `BlockingManager#blockingPublisherToVoidStage`, does ^ suggestion still apply with my latest changes?

wburns replied:
I suggested this originally and didn't go back to change this one. So using the new method is preferred.

ryanemerson commented:
@wburnsI have addressed everything except limiting the number of blocking operations.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,280 @@
+         }
+      }, "write-cache-config");
+
+      if (cache.isEmpty())
```
Technically this method can block as well. I would say we could use `Cache#sizeAsync` but that requires getting all elements. TBH, until we have a `Cache#isEmptyAsync` I would recommend just removing this if block. Although it would be quite simple to add this method to the cache if you wanted.

ryanemerson replied:
I'll remove the if block ... this PR is already pretty large

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -0,0 +1,280 @@
+                              })
+                              .doOnNext(e -> writeMessageStream(e, serCtx, output))
+                              .doOnError(t -> {
+                                 throw new CacheException("Unable to create cache backup", t);
```
Sorry I didn't notice this earlier, but this doesn't do what you want sadly.

You need to do something like https://github.com/infinispan/infinispan/blob/37abd44b44b51eb5b32b23f7f7ddb277f7661021/core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java#L378

wburns commented on core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:
```diff
@@ -246,6 +261,26 @@ protected void start() {
+      CompletionStage<Void> stage = Flowable.defer(() -> {
+         Flowable<V> flowable = Flowable.fromPublisher(publisher);
+         if (isCurrentThreadBlocking()) {
+            log.tracef("Invoked on a blocking thread, running %s in same blocking thread", traceId);
```
We can add an `if (trace)`

wburns commented on core/src/main/java/org/infinispan/util/concurrent/BlockingManager.java:
```diff
@@ -169,6 +184,19 @@
+    * <p>
+    * Note that if the current thread is blocking everything including subscription, publication and collection of
+    * values will be done on the current thread.
+    *
```
We should note the stage will always be completed upon a non blocking thread.

ryanemerson commented:
> LGTM but you have a test failure
> https://ci.infinispan.org/job/Infinispan/job/PR-8532/31/testReport/junit/org.infinispan.server.core.backup/BackupManagerImplTest/testBackupAndRestoreMultipleContainers/

This was caused by the removal of the `isEmpty` check on the cache. To keep things simple and avoid adding `isEmptyAsync`, an empty .dat file is created during backup for an empty cache. In the context of a backup operation this has a negligible impact on performance or backup size.

wburns commented:
Integrated into master, thanks @ryanemerson !
